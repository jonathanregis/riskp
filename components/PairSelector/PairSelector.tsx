import { useTokens } from "@/contexts/TokensContext";
import { pairSelectorProps } from "@/types/tokenPair";
import { ChangeEvent, useEffect, useState } from "react";

//I'm keeping this component simple. Normally I would create a custom select component or use a third party
//so I can render a better select dropdown with search capabilities and icons
//Also, I would normally make sure that you can't select the same token as base and quote at the same time. But time is running out.

export function PairSelector({ defaultBase = 'ETH', defaultQuote = 'USD', onApply }: pairSelectorProps) {
    const { tokens } = useTokens();
    const [base, setBase] = useState<string>(defaultBase);
    const [quote, setQuote] = useState<string>(defaultQuote);
    const change = (e: ChangeEvent<HTMLSelectElement>) => {
        if (e.target.name === "quote") {
            setQuote(e.target.value)
        } else {
            setBase(e.target.value)
        }
    }

    useEffect(() => {
        setBase(defaultBase);
        setQuote(defaultQuote);
    }, [defaultBase, defaultQuote])

    return <div className="flex flex-1 flex-wrap align-center justify-between">
        <div className="p-2">
            <p className="text-slate-400 uppercase text-xs">from</p>
            <select aria-label="from" onChange={change} value={base} name="base" className="border-none text-bold text-sm">
                {tokens.length && tokens.map((token) => {
                    return <option key={token.id} value={token.id}>{token.id} - {token.name}</option>
                })}
            </select>
        </div>
        <div className="p-2">
            <p className="text-slate-400 uppercase text-xs">to</p>
            <select aria-label="to" onChange={change} value={quote} name="quote" className="border-none text-bold text-sm">
                {tokens.length && tokens.map((token) => {
                    return <option key={token.id} value={token.id}>{token.id} - {token.name}</option>
                })}
            </select>
        </div>
        {(base != defaultBase || quote != defaultQuote) ? <button onClick={() => onApply(base, quote)} className="shadow-lg rounded-lg p-4 w-full flex-basis-full bg-[#daec03] text-bold uppercase m-4">update</button> : null}
    </div>
}