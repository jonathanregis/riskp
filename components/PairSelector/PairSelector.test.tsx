import { render, screen, fireEvent } from '@testing-library/react';
import { PairSelector } from './PairSelector';
import '@testing-library/jest-dom';

describe('PairSelector component', () => {
    const onApply = jest.fn();
    const defaultBase = 'ETH';
    const defaultQuote = 'USD';

    test('renders from and to select fields', () => {
        const { container } = render(<PairSelector defaultBase={defaultBase} defaultQuote={defaultQuote} onApply={onApply} />);
        const baseSelect = screen.getByLabelText('from');
        const quoteSelect = screen.getByLabelText('to');

        expect(baseSelect).toBeInTheDocument();
        expect(quoteSelect).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    test('calls onApply function when update button is clicked', () => {
        const { container } = render(<PairSelector defaultBase={defaultBase} defaultQuote={defaultQuote} onApply={onApply} />);
        const newBase = 'BTC';
        const newQuote = 'EUR';

        fireEvent.change(screen.getByLabelText('from'), { target: { value: newBase } });
        fireEvent.change(screen.getByLabelText('to'), { target: { value: newQuote } });
        setTimeout(() => {
            expect(screen.getByLabelText('from')).toHaveValue(newBase);
            expect(screen.getByLabelText('to')).toHaveValue(newQuote);
            const updateButton = screen.getByText('update');
            fireEvent.click(updateButton);

            expect(onApply).toHaveBeenCalledTimes(1);
            expect(onApply).toHaveBeenCalledWith(newBase, newQuote);
        }, 4000)

        expect(container).toMatchSnapshot();
    });
});
