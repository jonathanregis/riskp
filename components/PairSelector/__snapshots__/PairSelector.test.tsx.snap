// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PairSelector component calls onApply function when update button is clicked 1`] = `
<div>
  <div
    class="flex flex-1 flex-wrap align-center justify-between"
  >
    <div
      class="p-2"
    >
      <p
        class="text-slate-400 uppercase text-xs"
      >
        from
      </p>
      <select
        aria-label="from"
        class="border-none text-bold text-sm"
        name="base"
      >
        0
      </select>
    </div>
    <div
      class="p-2"
    >
      <p
        class="text-slate-400 uppercase text-xs"
      >
        to
      </p>
      <select
        aria-label="to"
        class="border-none text-bold text-sm"
        name="quote"
      >
        0
      </select>
    </div>
    <button
      class="shadow-lg rounded-lg p-4 w-full flex-basis-full bg-[#daec03] text-bold uppercase m-4"
    >
      update
    </button>
  </div>
</div>
`;

exports[`PairSelector component renders from and to select fields 1`] = `
<div>
  <div
    class="flex flex-1 flex-wrap align-center justify-between"
  >
    <div
      class="p-2"
    >
      <p
        class="text-slate-400 uppercase text-xs"
      >
        from
      </p>
      <select
        aria-label="from"
        class="border-none text-bold text-sm"
        name="base"
      >
        0
      </select>
    </div>
    <div
      class="p-2"
    >
      <p
        class="text-slate-400 uppercase text-xs"
      >
        to
      </p>
      <select
        aria-label="to"
        class="border-none text-bold text-sm"
        name="quote"
      >
        0
      </select>
    </div>
  </div>
</div>
`;
